### ISSUE TITLE
Incorrect logits computation in GRPOTrainer._get_per_token_logps_and_entropies

### ISSUE THREAD

**Issue by jhamrick** [huggingface/trl#3733]
### Reproduction

I believe that the way the per token logps is calculated in GRPOTrainer is incorrect due to incorrect attention masking. In the [code](https://github.com/huggingface/trl/blob/main/trl/trainer/grpo_trainer.py#L891), the logits are computed as:

```python
logits = model(
  input_ids=input_ids_batch,
  attention_mask=attention_mask_batch,
  logits_to_keep=logits_to_keep + 1,
).logits
```

where `input_ids_batch` and `attention_mask_batch` have shape `(generation_batch_size, prompt_length + completion_length)`. The attention mask is one for prompt tokens and zero for completion tokens. What this means is that the model is computing logits for each token *conditioned only on the prompt* rather than on all tokens earlier in the sequence.

The correct behavior should be something like:

```python
logits = []
for i in range(logits_to_keep):
    logits.append(model(
        input_ids=input_ids_batch[:, :(len(tokens) + i)],
        attention_mask=torch.ones((input_ids_batch.size(0), len(tokens) + i)),
        logits_to_keeps=logits_to_keep + 1,
    ).logits[:, -1])
logits = torch.stack(logits, dim=1)
```

Though this is not particularly efficient—I know you can ask for the logits to be returned in the `GenerationConfig` so this might be one way to do it but wouldn't help for computing logits from the reference model. The more efficient implementation would be I guess to break the computation up into batches with the learner's batch size.

Here is a Jupyter notebook demonstrating the issue, showing that the current behavior produces logits which are very different from those produced by the sampler, and which yield very low logp estimates for the probability of the sequence: https://gist.github.com/jhamrick/6ae2fab2f9e4719e88c6a95f65f6ff27

### System Info

- Platform: Linux-5.15.0-139-generic-x86_64-with-glibc2.31
- Python version: 3.10.13
- TRL version: 0.20.0.dev0
- PyTorch version: 2.7.0
- accelerator(s): NVIDIA H100 80GB HBM3
- Transformers version: 4.53.0
- Accelerate version: 1.8.1
- Accelerate config: not found
- Datasets version: 3.6.0
- HF Hub version: 0.33.2
- bitsandbytes version: 0.46.1
- DeepSpeed version: not installed
- Diffusers version: not installed
- Liger-Kernel version: not installed
- LLM-Blender version: not installed
- OpenAI version: 1.93.0
- PEFT version: 0.16.0
- vLLM version: 0.9.1

### Checklist

- [x] I have checked that my issue isn't already filed (see [open issues](https://github.com/huggingface/trl/issues?q=is%3Aissue))
- [x] I have included my system information
- [x] Any code provided is minimal, complete, and reproducible ([more on MREs](https://docs.github.com/en/get-started/writing-on-github/working-with-advanced-formatting/creating-and-highlighting-code-blocks))
- [x] Any code provided is properly formatted in code blocks, (no screenshot, [more on code blocks](https://docs.github.com/en/get-started/writing-on-github/working-with-advanced-formatting/creating-and-highlighting-code-blocks))
- [x] Any traceback provided is complete

**Comment by jhamrick** [comment]
Hm, digging into this some more I think because I had set `mask_truncated_completions=True`, it was causing the completion mask to be zeroed out and thus leading to wrong probability calculations (plus me misunderstanding the attention_mask argument). I think the fact that we still get logps out in this case is not what I'd expect, but the issue isn't what I thought originally, so I'll close this.

**Comment by gau-nernst** [comment]
@jhamrick I'm digging thru verifiers and TRL code. From my understanding, shouldn't `attention_mask` NOT be passed to the model when computing logits? The mask should only be used for per-token loss masking, not for attention. If we pass `attention_mask` to the model, doesn't it mean we **ignore the prompt, only condition on the response**? Which is not right.

**Comment by jhamrick** [comment]
AFAIU it's a bit of a misnomer—it's not really a $N\times N$-shaped attention mask but a $N$-shaped token mask.

Assuming `mask_truncated_completions=False` (which was what had confused me), then `attention_mask_batch =  torch.ones((generation_batch_size, prompt_length + completion_length))`. Therefore every token will be attended to subject to causal attention constraints which happen deeper in the code.

When `mask_truncated_completions=True`, then `attention_mask_batch =  torch.cat([torch.ones((generation_batch_size, prompt_length)), torch.zeros((generation_batch_size, completion_length))], dim=1)`. This means that the logits returned for the completion tokens are effectively garbage (they ignore the *response* and condition only on the prompt), but it's ok because they're masked out later in the loss anyway (with `completion_mask`).

**Comment by gau-nernst** [comment]
I was reviewing this part again. Apparently verifiers implements this differently so my problem was different. It looks like there is no problem with TRL.

Anyway, I was thinking, if we limit to causal attention, passing attention mask to the model is not necessary. Per-token loss is masked out later anyway.
